File: consume/pecsplit.js
#!/usr/bin/nodejs

// Parsing the variable

var unquote = function(x) {
  if (x.search(/%[^#+.]|%$/) >= 0)
    throw new Error("invalid value: bad escape: " + x);
  return x.replace(/%\./g, ':').replace(/%\+/g, '=').replace(/%#/g, '%');
};

var parse_prefix_map = function(x) {
  return (x || "").split(/:/g).filter(Boolean).map(function(part) {
    var tuples = part.split(/=/g).map(unquote);
    if (tuples.length !== 2)
      throw new Error("invalid value: not a pair: " + tuples);
    return tuples;
  });
};

// Applying the variable

var map_prefix = function(string, pm) {
  for (var i = pm.length - 1; i >= 0; --i) {;
    var dst = pm[i][0];
    var src = pm[i][1];
    if (string.indexOf(src) === 0) {
      return dst + string.substr(src.length);
    }
  }
  return string;
};

// Main program

try {
  var pm = parse_prefix_map(process.env["BUILD_PATH_PREFIX_MAP"]);
  // var i = 2 is just how nodejs yolos its way through common conventions
  for (var i = 2, l = process.argv.length; i < l; ++i) {
    console.log(map_prefix(process.argv[i], pm));
  }
} catch (e) {
  console.error(e.stack);
  process.exit(1); // older versions returned 8, messing up our CI testing
}

File: consume/pecsplit.py
#!/usr/bin/python3

import re
import os
import sys

# Parsing the variable

def _dequote(part):
  if re.search(r"%[^pec]|%$", part):
    raise ValueError("bad escape: %s" % part)
  return part.replace("%c", ':').replace("%e", '=').replace("%p", '%');

def decode(prefix_str):
  tuples = (part.split("=") for part in prefix_str.split(":") if part)
  # Will raise if any tuple can't be destructured into a pair
  return [(_dequote(src), _dequote(dst)) for src, dst in tuples]

# Applying the variable

def map_prefix(string, pm):
  for src, dst in reversed(pm):
    if string.startswith(src):
      return dst + string[len(src):]
  return string

# Main program

pm = decode(os.getenv("BUILD_PATH_PREFIX_MAP", ""))
for v in sys.argv[1:]:
  # print() tries to auto-encode its args without surrogateescape
  sys.stdout.buffer.write(map_prefix(v, pm).encode("utf-8", errors="surrogateescape") + b"\n")
